### ISSUE TITLE
[Contenido - I3] duda conceptual tabla de páginas

### ISSUE THREAD

**Issue by rodrigooflores** [IIC2343/Syllabus-2024-2#146]
Buenas tardes, resolviendo algunas interrogaciones pasadas me surgió una pequeña confusión acerca del funcionamiento de la tabla de páginas y en qué situaciones se puede o hay que asignar marcos a la tabla de páginas.  

Por ejemplo en la I2-2023-2 pregunta 3 b)
<img width="500" alt="I2-2023-2_pregunta3b" src="https://github.com/user-attachments/assets/d15ee55a-6ee2-4b1c-81b5-d448d45e7ed5">
se va accediendo a direcciones virtuales y revisando si hay un marco asociado en la tabla de páginas _(para poder obtener la dirección física)_, y entiendo que el recorrido que se hace desde que se accede a la dirección virtual hasta que se obtiene (o quizás no) la dirección física es más o menos de la siguiente manera:
<img width="986" alt="esquema_acceso_tabla_pagina" src="https://github.com/user-attachments/assets/14a4477f-30bf-43de-888b-28b4b69ebed6">
pero en la I2-2022-2 pregunta 2 c)
<img width="500" alt="I2-2022-2_pregunta2c" src="https://github.com/user-attachments/assets/3bf6db7b-d4f7-4aab-827a-8de386bad526">
lo que se hace es que cada vez que ocurre un page fault se le asigna un marco según la política de reemplazo LRU, pero en este caso no me queda muy claro cómo quedaría el recorrido. ¿Lo que se está haciendo en este último caso es ir a buscar al disco en el swap file los marcos, y por eso podemos asignarle marcos por nuestra cuenta a la tabla de páginas? ¿es porque nos dan una política de reemplazo que podemos asignar marcos a la tabla de páginas en este caso?

**Comment by Geeermy** [comment]
¡Hola!

Por temas pedagógicos, asumimos que en general un _page fault_ implica que el contenido de la página está en el _swap file_ y no en un marco físico. Entonces sí, en el ejercicio, cada _page fault_ implica una búsqueda del contenido en el _swap file_ y posteriormente el traspaso (_swap in_) del contenido a un marco físico, ya sea el primero disponible o uno escogido según la política de reemplazo (en este caso, LRU, lo que implica además el _swap out_ del contenido del marco a ser reemplazado hacia el _swap file_).

Avísame si te queda claro.

¡Saludos!

**Comment by rodrigooflores** [comment]
Creo que sí entiendo, entonces si en el ejercicio de la I2-2023-2 si nos dieran un timestamp y nos pidieran política de reemplazo LRU ¿podríamos asignar la página 4 al marco 10b = 2 y la página 0 al marco 11b = 3 por ejemplo?

**Comment by Geeermy** [comment]
@rodrigooflores Claro, dependería del _timestamp_. Ahora, en el ejercicio la elección no es arbitraria, se puede deducir: Como partes con los cuatro marcos físicos disponibles, ninguno tiene _timestamp_ válido, pero eso va cambiando a medida que se realizan las traducciones. Dejaré una tabla con _timestamps_ a continuación.

Iteración 0 (Estado inicial)
Marco 0 | _Timestamp_ = -1
Marco 1 | _Timestamp_ = -1
Marco 2 | _Timestamp_ = -1
Marco 3 | _Timestamp_ = -1

Iteración 1 - Se asigna el marco físico 0, por lo que se actualiza su _timestamp_
**Marco 0 | _Timestamp_ = 0**
Marco 1 | _Timestamp_ = -1
Marco 2 | _Timestamp_ = -1
Marco 3 | _Timestamp_ = -1

Iteración 2 - Se asigna el marco físico 1, por lo que se actualiza su _timestamp_
Marco 0 | _Timestamp_ = 1 **(+1 al tiempo)**
**Marco 1 | _Timestamp_ = 0**
Marco 2 | _Timestamp_ = -1
Marco 3 | _Timestamp_ = -1

Iteración 3 - Se accede al marco físico 0 (_hit_), por lo que se actualiza su _timestamp_
**Marco 0 | _Timestamp_ = 0**
Marco 1 | _Timestamp_ = 1 **(+1 al tiempo)**
Marco 2 | _Timestamp_ = -1
Marco 3 | _Timestamp_ = -1

Iteración 4 - Se asigna el marco físico 2, por lo que se actualiza su _timestamp_
Marco 0 | _Timestamp_ = 1 **(+1 al tiempo)**
Marco 1 | _Timestamp_ = 2 **(+1 al tiempo)**
**Marco 2 | _Timestamp_ = 0**
Marco 3 | _Timestamp_ = -1

Iteración 5 - Se asigna el marco físico 3, por lo que se actualiza su _timestamp_
Marco 0 | _Timestamp_ = 2 **(+1 al tiempo)**
Marco 1 | _Timestamp_ = 3 **(+1 al tiempo)**
Marco 2 | _Timestamp_ = 1 **(+1 al tiempo)**
**Marco 3 | _Timestamp_ = 0**

Iteración 6 - Se debe asignar un marco físico nuevo. Por _timestamp_, se escoge el marco físico 1. Este es ahora el más reciente y se actualizan todos los _timestamps_.
.Marco 0 | _Timestamp_ = 3 **(+1 al tiempo)**
**Marco 1 | _Timestamp_ = 0**
Marco 2 | _Timestamp_ = 2 **(+1 al tiempo)**
Marco 3 | _Timestamp_ = 1 **(+1 al tiempo)**

¿Se entiende la idea? En este caso, si bien lo llamo _timestamp_, este no es más que es un número que me permite mantener el orden en el cual han sido accedidos los marcos, para así seleccionar el último en caso de requerir un reemplazo. Solo sabiendo que todos los marcos físicos estaban disponibles, se pudo deducir el marco físico a reemplazar cuando llegó el momento de hacerlo.

Espero que ahora haya quedado más claro. :)

¡Saludos!

**Comment by Geeermy** [comment]
**Para complementar:** Se hace distinto en el ejercicio del 2023-2 porque aquí no te piden llevar a cabo la traducción en caso de _page fault_, solo indicar que hubo uno. Entonces, no es necesario que hagas reemplazos de marco físico ni nada similar. En ese sentido, en el ejercicio del 2023-2 se trabaja con una tabla de páginas estáticas, no así en el del 2022-2.

**Comment by rodrigooflores** [comment]
Me quedó súper claro, muchísimas gracias :)
Saludos!
